TESTS.md

```diff
@@ -0,0 +1,138 @@
+
+    <property name="chameleonTarget">wildfly:10.0.0.Final:managed</property>
+
+Any other application server and its version can be used (such as EAP7), refer to Arquillian Chameleon documentation.
+
+Remote run mode
+---------------
+
+If you want to run/debug test against remote container (that speeds up execution), you have to start required containers first - right now two containers are required with certain post-offset:
+
+    ./bin/standalone.sh -c standalone-full-ha.xml -Djboss.socket.binding.port-offset=4321 -Djboss.node.name=node1 -Djboss.messaging.cluster.password=somepassword -Djava.net.preferIPv4Stack=true
+    ./bin/standalone.sh -c standalone-full-ha.xml -Djboss.socket.binding.port-offset=4322 -Djboss.node.name=node2 -Djboss.messaging.cluster.password=somepassword -Djava.net.preferIPv4Stack=true
+
+Since Arquillian is configured to detect running container (`allowConnectingToRunningServer=true`), it will detect running container and you don't have to anything else to run tests in dedicated test server:
+
+    mvn verify -Dtest=TestMetricCollector
+
+Debugging tests in Eclipse
+--------------------------
+
+Since Arquillian tests run in application server, it may be harder to debug them, but with IDE, it is quite easy:
+
+1. start Eclipse
+2. start two WildFly servers in Debug mode, using `standalone-full-ha.xml` and configured `port-offset` (see the configuration properties above)
+3. insert breakpoint into a test or an implementation code
+4. run the test class (or just selected method) by selecting its name in the IDE and choosing `Run as... JUnit Test`
+5. the breakpoint will be hit once server deploys micro-deployment and executes the test method
+
+If you need to debug failing test, it may be helpful to reveal what is going on in Arquillian - turn on its debug mode by passing following VM argument to the container (in Eclipse you need to modify launch configuration):
+
+    -Darquillian.debug=true
+
+Additionally, turn on AeroGear debug log by executing following CLI commands:
+
+    /subsystem=logging/console-handler=AEROGEAR:add(formatter="%d{HH:mm:ss,SSS} %-5p [%c] (%t) %s%E%n",autoflush=true)
+    /subsystem=logging/logger=org.jboss.aerogear.unifiedpush:add(level=FINEST,use-parent-handlers=false,handlers=["AEROGEAR"])
+
+Writing new tests
+-----------------
+
+Since tests are deployed to running container, they can reuse anything that is deployed with them:
+
+* inject CDI (`@Inject`) or EJB (`@EJB`) beans
+* inject resources such as data sources or JMS destinations (`@Resource`)
+* mock dependencies using Mockito (see `UnifiedPushArchive#withMockito`)
+
+Tests can leverage basic JUnit features such as `@Before` and `@After`, `@Test(timeout)`, `@Test(expected)` or even `@Rule`, etc.
+
+Tests for cluster
```

Comment:
@lfryc should some info about multicast on OSX in here - or a link to https://github.com/aerogear/aerogear-unifiedpush-server#getting-started-with-clustered-servers ?